diff -Naur wine-1.1.19-orig/dlls/wined3d/baseshader.c wine-1.1.19-final/dlls/wined3d/baseshader.c
--- wine-1.1.19-orig/dlls/wined3d/baseshader.c	2009-04-12 23:04:02.023062425 +0000
+++ wine-1.1.19-final/dlls/wined3d/baseshader.c	2009-04-12 23:04:29.090143505 +0000
@@ -370,6 +370,11 @@
             pToken += curOpcode->num_params;
 
             /* Rep and Loop always use an integer constant for the control parameters */
+	    if (!This->baseShader.uses_int_consts && !pshader) {
+		/* need to reserve room for the int uniforms */
+		This->baseShader.limits.constant_float -= MAX_CONST_I;
+		TRACE_(d3d_shader)("Reserving int uniforms, adjusting constant_float for this vshader.\n");
+	    }
             This->baseShader.uses_int_consts = TRUE;
         } else if (WINED3DSIO_ENDLOOP == curOpcode->opcode ||
                    WINED3DSIO_ENDREP == curOpcode->opcode) {
@@ -492,6 +497,11 @@
                     }
                 }
                 else if(WINED3DSPR_CONSTINT == regtype) {
+		   if (!This->baseShader.uses_int_consts && !pshader) {
+			/* need to reserve room for the int uniforms */
+			This->baseShader.limits.constant_float -= MAX_CONST_I;
+			TRACE_(d3d_shader)("Reserving int uniforms, adjusting constant_float for this vshader.\n");
+		    }
                     This->baseShader.uses_int_consts = TRUE;
                 }
                 else if(WINED3DSPR_CONSTBOOL == regtype) {
diff -Naur wine-1.1.19-orig/dlls/wined3d/directx.c wine-1.1.19-final/dlls/wined3d/directx.c
--- wine-1.1.19-orig/dlls/wined3d/directx.c	2009-04-12 23:04:02.026396179 +0000
+++ wine-1.1.19-final/dlls/wined3d/directx.c	2009-04-12 23:04:29.093476561 +0000
@@ -956,14 +956,17 @@
         }
         if (gl_info->supported[ARB_VERTEX_SHADER]) {
             glGetIntegerv(GL_MAX_VERTEX_UNIFORM_COMPONENTS_ARB, &gl_max);
+            gl_max = wined3d_settings.max_vertex_uniforms_override ? wined3d_settings.max_vertex_uniforms_override : gl_max;
             gl_info->vs_glsl_constantsF = gl_max / 4;
             TRACE_(d3d_caps)("Max ARB_VERTEX_SHADER float constants: %u\n", gl_info->vs_glsl_constantsF);
         }
         if (gl_info->supported[ARB_FRAGMENT_SHADER]) {
             glGetIntegerv(GL_MAX_FRAGMENT_UNIFORM_COMPONENTS_ARB, &gl_max);
+            gl_max = wined3d_settings.max_fragment_uniforms_override ? wined3d_settings.max_fragment_uniforms_override : gl_max;
             gl_info->ps_glsl_constantsF = gl_max / 4;
             TRACE_(d3d_caps)("Max ARB_FRAGMENT_SHADER float constants: %u\n", gl_info->ps_glsl_constantsF);
             glGetIntegerv(GL_MAX_VARYING_FLOATS_ARB, &gl_max);
+            gl_max = wined3d_settings.max_varying_override ? wined3d_settings.max_varying_override : gl_max;
             gl_info->max_glsl_varyings = gl_max;
             TRACE_(d3d_caps)("Max GLSL varyings: %u (%u 4 component varyings)\n", gl_max, gl_max / 4);
         }
diff -Naur wine-1.1.19-orig/dlls/wined3d/glsl_shader.c wine-1.1.19-final/dlls/wined3d/glsl_shader.c
--- wine-1.1.19-orig/dlls/wined3d/glsl_shader.c	2009-04-12 23:04:02.026396179 +0000
+++ wine-1.1.19-final/dlls/wined3d/glsl_shader.c	2009-04-12 23:06:07.266800392 +0000
@@ -728,7 +728,7 @@
             max_constantsF = min(This->baseShader.limits.constant_float, max_constantsF);
         } else {
             /* Subtract the other potential uniforms from the max available (bools, ints, and 1 row of projection matrix) */
-            max_constantsF = GL_LIMITS(vshader_constantsF) - (MAX_CONST_B / 4) - MAX_CONST_I - 1;
+            max_constantsF = GL_LIMITS(vshader_constantsF) - (MAX_CONST_B / 4) - 1;
             max_constantsF = min(This->baseShader.limits.constant_float, max_constantsF);
         }
         shader_addline(buffer, "uniform vec4 %cC[%u];\n", prefix, max_constantsF);
@@ -4150,7 +4150,7 @@
         pCaps->VertexShaderVersion = WINED3DVS_VERSION(3,0);
     TRACE_(d3d_caps)("Hardware vertex shader version %d.%d enabled (GLSL)\n", (pCaps->VertexShaderVersion >> 8) & 0xff, pCaps->VertexShaderVersion & 0xff);
     /* Subtract the other potential uniforms from the max available (bools, ints, and 1 row of projection matrix) */
-    pCaps->MaxVertexShaderConst = GL_LIMITS(vshader_constantsF) - (MAX_CONST_B / 4) - MAX_CONST_I - 1;
+    pCaps->MaxVertexShaderConst = GL_LIMITS(vshader_constantsF) - (MAX_CONST_B / 4) - 1;
 
     /* Older DX9-class videocards (GeforceFX / Radeon >9500/X*00) only support pixel shader 2.0/2.0a/2.0b.
      * In OpenGL the extensions related to GLSL abstract lowlevel GL info away which is needed
diff -Naur wine-1.1.19-orig/dlls/wined3d/wined3d_main.c wine-1.1.19-final/dlls/wined3d/wined3d_main.c
--- wine-1.1.19-orig/dlls/wined3d/wined3d_main.c	2009-04-12 23:04:02.033062708 +0000
+++ wine-1.1.19-final/dlls/wined3d/wined3d_main.c	2009-04-12 23:04:29.093476561 +0000
@@ -304,6 +304,33 @@
                 wined3d_settings.allow_multisampling = TRUE;
             }
         }
+        if ( !get_config_key( hkey, appkey, "MaxVertexUniforms", buffer, size) )
+        {
+            int TmpMaxVerts = atoi(buffer);
+            if(TmpMaxVerts > 0)
+            {
+                TRACE("Overriding glxinfo for GL_MAX_VERTEX_UNIFORM_COMPONENTS_ARB now set to %i\n", TmpMaxVerts);
+                wined3d_settings.max_vertex_uniforms_override = TmpMaxVerts;
+            }
+        }
+        if ( !get_config_key( hkey, appkey, "MaxFragmentUniforms", buffer, size) )
+        {
+            int TmpMaxFrags = atoi(buffer);
+            if(TmpMaxFrags > 0)
+            {
+                TRACE("Overriding glxinfo for GL_MAX_FRAGMENT_UNIFORM_COMPONENTS_ARB now set to %i\n", TmpMaxFrags);
+                wined3d_settings.max_fragment_uniforms_override = TmpMaxFrags;
+            }
+        }
+        if ( !get_config_key( hkey, appkey, "MaxVaryings", buffer, size) )
+        {
+            int TmpMaxVarying = atoi(buffer);
+            if(TmpMaxVarying > 0)
+            {
+                TRACE("Overriding glxinfo for GL_MAX_VARYING_FLOATS_ARB now set to %i\n", TmpMaxVarying);
+                wined3d_settings.max_varying_override = TmpMaxVarying;
+            }
+        }
     }
     if (wined3d_settings.vs_mode == VS_HW)
         TRACE("Allow HW vertex shaders\n");
diff -Naur wine-1.1.19-orig/dlls/wined3d/wined3d_private.h wine-1.1.19-final/dlls/wined3d/wined3d_private.h
--- wine-1.1.19-orig/dlls/wined3d/wined3d_private.h	2009-04-12 23:04:02.033062708 +0000
+++ wine-1.1.19-final/dlls/wined3d/wined3d_private.h	2009-04-12 23:04:29.093476561 +0000
@@ -290,6 +290,10 @@
   unsigned int emulated_textureram;
   char *logo;
   int allow_multisampling;
+/* overrides for GL_MAX_* parameters */
+  int max_vertex_uniforms_override;
+  int max_fragment_uniforms_override;
+  int max_varying_override;
 } wined3d_settings_t;
 
 extern wined3d_settings_t wined3d_settings;
